### ISSUE TITLE
[workspaces] assertion triggered when processing build requirements

### ISSUE THREAD

**Issue by theodelrieu** [conan-io/conan#5612]
Hello,

This one is quite hard to reproduce minimally, I'll try my best to help debug it.
Here is the stacktrace I get when running `conan workspace install`:

>Traceback (most recent call last):
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/command.py", line 1859, in run
    method(args[0][1:])
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/command.py", line 1661, in workspace
    install_folder=args.install_folder)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/conan_api.py", line 78, in wrapper
    return f(*args, **kwargs)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/conan_api.py", line 480, in workspace_install
    False, update, remotes, recorder)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 182, in load_graph
    graph_lock=graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 284, in _load_graph
    apply_build_requires=apply_build_requires)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 255, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_manager.py", line 251, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 65, in extend_build_requires
    remotes, processed_profile, new_reqs, new_options, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 178, in _handle_require
    update, remotes, processed_profile, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 132, in _load_deps
    remotes, processed_profile, new_reqs, new_options, graph_lock)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph_builder.py", line 200, in _handle_require
    dep_graph.add_edge(node, previous, require)
  File "/home/theo/.local/lib/python3.7/site-packages/conans/client/graph/graph.py", line 213, in add_edge
    assert src in self.nodes and dst in self.nodes
AssertionError

Here is the conanfile.py causing trouble:

```python
from conans import python_requires


base = python_requires("theo_conanfile/dev")

class BugConan(base.get_conanfile()):
    name = "bug"

    def build_requirements(self):
        if self.should_build_tests:
            # commenting the following line avoids the assert
            self.build_requires("functional-helpers/%s@theo/testing" % self.version)
        # The following adds 3 build requirements, but functional-helpers is not one of them
        super().build_requirements()

    def requirements(self):
        self.requires("log/%s@theo/testing" % self.version)
        self.requires("errors/%s@theo/testing" % self.version)

        self.requires("fmt/5.3.0@bincrafters/stable")
        self.requires("jsonformoderncpp/3.4.0@theo/testing")
        self.requires("optional-lite/3.1.1@theo/testing")
```

I'm using Conan 1.8.1, with revisions enabled.

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [ ] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by jgsogo** [comment]
Hi, @theodelrieu I'll try to reproduce the issue, but I'll need some help:
 * Does it happen with a regular install too or only inside a workspace?
 * Does it have anything to do with the reference without user/channel you are using in the `python_requires`? Does it fails with a full reference?
 * Can you share with us the source of `theo_conanfile/dev` so we can know requires of that base class and the implementation around `should_build_tests`?

Thanks!

**Comment by theodelrieu** [comment]
Hi @jgsogo,

I should have posted everything in the OP...

>Does it happen with a regular install too or only inside a workspace?

I cannot really tell, since I'm splitting the code while moving to workspaces

>Does it have anything to do with the reference without user/channel you are using in the python_requires? Does it fails with a full reference?

I just tested, it also fails with a full reference.

Here is the implementation of `theo_conanfile/dev`:

```python
from conans import ConanFile, CMake, tools


class PythonRequires(ConanFile):
    name = "theo_conanfile"
    version = "dev"

def get_conanfile(additional_options = dict(), additional_default_options = dict()):
    class BaseConanFile(ConanFile):
        settings = "os", "compiler", "build_type", "arch"
        options = {
                "shared": [True, False],
                "fPIC": [True, False],
                "warn_as_error": [True, False],
                "sanitizer": ["address", "leak", "memory", "thread", "undefined", None],
                "coroutinests": [True, False]
            }
        options.update(additional_options)
        default_options = {
                "shared": False,
                "fPIC": True,
                "warn_as_error": False,
                "sanitizer": None,
                "coroutinests": False
            }
        default_options.update(additional_default_options)

        generators = "cmake"
        exports_sources = "src/*", "include/*", "CMakeLists.txt", "test/*"
        cmake = None


        @property
        def sanitizer_flag(self):
            if self.options.sanitizer:
                return " -fsanitize=%s " % self.options.sanitizer
            return None

        @property
        def cross_building(self):
            return tools.cross_building(self.settings)

        @property
        def should_build_tests(self):
            # develop is false when the package is used as a requirement.
            return self.develop and (not self.cross_building or self.settings.os == "Emscripten")

        def init_cmake(self):
            if self.cmake:
                return

            self.cmake = CMake(self)

            if "CONAN_CXX_FLAGS" not in self.cmake.definitions:
                self.cmake.definitions["CONAN_CXX_FLAGS"] = ""
            if "CONAN_C_FLAGS" not in self.cmake.definitions:
                self.cmake.definitions["CONAN_C_FLAGS"] = ""

            if self.options.sanitizer:
                self.cmake.definitions["CONAN_C_FLAGS"] += self.sanitizer_flag
                self.cmake.definitions["CONAN_CXX_FLAGS"] += self.sanitizer_flag
            if self.options.coroutinests:
                self.cmake.definitions["CONAN_CXX_FLAGS"] += " -fcoroutines-ts "
            self.cmake.definitions["BUILD_TESTING"] = self.should_build_tests

        def configure(self):
            pass

        def config_options(self):
            pass

        def build_requirements(self):
            if self.should_build_tests:
                self.build_requires("docopt.cpp/0.6.2@theo/testing")
                self.build_requires("doctest/2.2.3@theo/testing")
                self.build_requires("doctest-async/2.0.12@theo/testing")

        def requirements(self):
            pass

        def build(self):
            self.init_cmake()
            self.cmake.configure()
            self.cmake.build()
            self.cmake.definitions["BUILD_TESTING"] = self.should_build_tests
            self.cmake.definitions["WARN_AS_ERROR"] = self.options.warn_as_error

        def package(self):
            self.init_cmake()
            self.cmake.install()

        def package_info(self):
            if self.sanitizer_flag:
                self.cpp_info.sharedlinkflags = [self.sanitizer_flag]
                self.cpp_info.exelinkflags = [self.sanitizer_flag]
            self.cpp_info.libs = [self.name]

        def package_id(self):
            del self.info.options.warn_as_error

        def build_id(self):
            pass

    return BaseConanFile
```

**Comment by jgsogo** [comment]
I'm trying to reproduce your issue with the files in this repo (https://github.com/jgsogo/5612-ws-issue), but without success. I've been trying to simplify the conanfile to get rid of everything related to building stuff and not used dependencies... 

I need more help to reproduce the issue 😞 , feel free to comment there or even to modify the repo directly (I've added you as collaborator).

Thanks

**Comment by theodelrieu** [comment]
Thanks, I'll try to reproduce it

**Comment by jgsogo** [comment]
Thanks to @theodelrieu I have a minimum reproducible example with the error: https://github.com/jgsogo/5612-ws-issue/tree/theo-min

Modifying a bit the Conan code, we can get more information for the error:

```bash
    assert dst in self.nodes, "Cannot find dst='{}' in nodes (src='{}')".format(dst, src)
AssertionError: Cannot find dst='fmt/5.3.0@bincrafters/stable' in nodes (src='theo_test-helpers/dev@theo/testing')

ERROR: Cannot find dst='fmt/5.3.0@bincrafters/stable' in nodes (src='theo_test-helpers/dev@theo/testing')
```

I'll add a test reproducing the issue.

**Comment by jgsogo** [comment]
PR https://github.com/conan-io/conan/pull/5622 reproduces this issue

**Comment by jgsogo** [comment]
Closed via https://github.com/conan-io/conan/issues/5689

**Comment by jgsogo** [comment]
This is already fixed (and released) in Conan `1.18.2`

**Comment by theodelrieu** [comment]
\o/
